### PR TITLE
Peer filter now takes a lookup interface

### LINKED ISSUE #1274: [Core] Rename ValidatorAddress
We want to better indicate what a node's core signing key means - possibly `SigningKeyAddress`.

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill greg.hill@monax.io

Revert `Address` -> `ValidatorAddress`
Closes #1274

Faster peer lookup for identifyTx
lookup by nodeID instead

### CODE REVIEW

**Comment by gregdhill** [comment]
This is a rough improvement on the functionality we had before so instead of looping over all available ids or addresses we'll just do an assertion. This refactor did uncover one nuance with identify in that new nodes must always allow an initial connection. Another slightly finicky issue is address lookup, since we don't impose any unique constraint we have to return a list of possible nodes registered with that net address. Currently this is just a mapping, but I wonder if we should store this in the `plain` as it has the possibility of growing quite large.

File: execution/registry/cache.go [hunk 1]
```diff
@@ -106,22 +107,22 @@ func (cache *Cache) RemoveNode(addr crypto.Address) error {
...
-		addresses = append(addresses, addr)
+	var ids []crypto.Address
+	for id := range cache.registry {
+		ids = append(ids, id)
 	}
 
-	for _, addr := range addresses {
-		info := cache.registry[addr]
+	for _, id := range ids {
+		info := cache.registry[id]
```
Comment (seanyoung) [inline execution/registry/cache.go]: Here you're building an array just to loop over it. Could it not be simplified with:
for info, id := range cache.registry {
}
and do away with ids?
note that if you know the size of an array, best to allocate it with that capicity to reprevent reallocs:
ids := make([]crypto.Address, 0, len(cache.registry))
however I don't think ids is needed.

File: execution/registry/cache.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ type Cache struct {
 	sync.RWMutex
 	backend  Reader
 	registry map[crypto.Address]*nodeInfo
+	stats    NodeStats
 }
 
 type nodeInfo struct {
```
Comment (silasdavis) [inline execution/registry/cache.go]: SUCHCACHEMUCHWOW

File: execution/registry/registry.go [hunk 3]
```diff
@@ -16,38 +16,112 @@ package registry
...
 )
 
-func AuthorizedPeersProvider(state IterableReader) func() ([]string, []string) {
-	return func() ([]string, []string) {
-		var peerIDs, peerAddrs []string
+type NodeStats struct {
+	Addresses map[string]map[crypto.Address]struct{}
+}
+
+func NewNodeStats() NodeStats {
```
Comment (silasdavis) [inline execution/registry/registry.go]: On the one hand I like that it rhymes with `AccountStats` and has the same lifecycle on the other hand it's not really stats... probably fine.

File: execution/state/state.go [hunk 4]
```diff
@@ -215,7 +215,7 @@ func LoadState(db dbm.DB, version int64) (*State, error) {
 		return nil, err
 	}
 
-	err = s.loadNodeList()
+	err = s.loadNodeStats()
```
Comment (silasdavis) [inline execution/state/state.go]: Can we have a comment saying why we are doing this.

File: execution/registry/registry.go [hunk 5]
```diff
@@ -16,38 +16,112 @@ package registry
...
+	}
+	return addrs
+}
+
+func (ns *NodeStats) Insert(net string, id crypto.Address) {
+	_, ok := ns.Addresses[net]
+	if !ok {
+		ns.Addresses[net] = make(map[crypto.Address]struct{})
+	}
+	ns.Addresses[net][id] = struct{}{}
```
Comment (silasdavis) [inline execution/registry/registry.go]: `Address` is very overloaded, in some places we have `NodeAddress`. Can we use `NetworkAddress` or `NetAddress` to make it clearer when we are talking about network.

**Pull request merged**